### PR TITLE
KAFKA-14953: Add tiered storage related metrics

### PR DESCRIPTION
**Pull request by abhijeetk88**
In this PR, I have added the following metrics that are related to tiered storage mentioned in[ KIP-405](https://cwiki.apache.org/confluence/display/KAFKA/KIP-405%3A+Kafka+Tiered+Storage).
|Metric|Description|
|------|------------|
| RemoteReadRequestsPerSec| Number of remote storage read requests per second|
| RemoteWriteRequestsPerSec| Number of remote storage write requests per second|
| RemoteBytesInPerSec|Number of bytes read from remote storage per second|
| RemoteReadErrorsPerSec|Number of remote storage read errors per second|
| RemoteBytesOutPerSec|Number of bytes copied to remote storage per second|
| RemoteWriteErrorsPerSec|Number of remote storage write errors per second|
| RemoteLogReaderTaskQueueSize|Number of remote storage read tasks pending for execution|
| RemoteLogReaderAvgIdlePercent| Average idle percent of the remote storage reader thread pool|
| RemoteLogManagerTasksAvgIdlePercent|Average idle percent of RemoteLogManager thread pool|

Added unit tests for all the rate metrics.

I will have a follow-up PR to add these metrics to Kafka documentation.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by showuon** [comment]
@abhijeetk88 , please let me know when PR is ready for review. Thanks for the work on weekend! :)

**Comment by abhijeetk88** [comment]
@showuon The PR is ready for review. Please take a look.

**Comment by showuon** [comment]
Thanks @abhijeetk88 , will review it later this week.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java [hunk 1]
```diff
@@ -158,23 +164,34 @@ public RemoteLogManager(RemoteLogManagerConfig rlmConfig,
...
         this.fetchLog = fetchLog;
+        this.brokerTopicStats = brokerTopicStats;
 
         remoteLogStorageManager = createRemoteStorageManager();
         remoteLogMetadataManager = createRemoteLogMetadataManager();
         indexCache = new RemoteIndexCache(1024, remoteLogStorageManager, logDir);
         delayInMs = rlmConfig.remoteLogManagerTaskIntervalMs();
         rlmScheduledThreadPool = new RLMScheduledThreadPool(rlmConfig.remoteLogManagerThreadPoolSize());
+
+        metricsGroup.newGauge("RemoteLogManagerTasksAvgIdlePercent", new Gauge<Double>() {
```
Comment (divijvaidya) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java]: we need to close the gauge on RLM shutdown. You can see PRs attached to https://issues.apache.org/jira/browse/KAFKA-15129 as examples.

Also, we need to add tests to validate that they are close, move metric names into constants etc.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 2]
```diff
@@ -277,6 +277,11 @@ class BrokerTopicMetrics(name: Option[String]) {
     BrokerTopicStats.TotalFetchRequestsPerSec -> MeterWrapper(BrokerTopicStats.TotalFetchRequestsPerSec, "requests"),
     BrokerTopicStats.FetchMessageConversionsPerSec -> MeterWrapper(BrokerTopicStats.FetchMessageConversionsPerSec, "requests"),
     BrokerTopicStats.ProduceMessageConversionsPerSec -> MeterWrapper(BrokerTopicStats.ProduceMessageConversionsPerSec, "requests"),
+    BrokerTopicStats.RemoteBytesOutPerSec -> MeterWrapper(BrokerTopicStats.RemoteBytesOutPerSec, "bytes"),
```
Comment (divijvaidya) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala:280]: We should probably add Remote metrics only when RemoteStorage is enabled on the cluster. Otherwise these these per-topic metrics are useless.

File: core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java [hunk 3]
```diff
@@ -61,15 +74,20 @@ public void testRemoteLogReaderWithoutError() throws RemoteStorageException, IOE
...
+        // Verify metrics for remote reads are updated correctly
+        assertEquals(1, brokerTopicStats.topicStats(TOPIC).remoteReadRequestRate().count());
+        assertEquals(100, brokerTopicStats.topicStats(TOPIC).remoteBytesInRate().count());
+        assertEquals(0, brokerTopicStats.topicStats(TOPIC).failedRemoteReadRequestRate().count());
     }
 
     @Test
     public void testRemoteLogReaderWithError() throws RemoteStorageException, IOException {
-        when(mockRLM.read(any(RemoteStorageFetchInfo.class))).thenThrow(new OffsetOutOfRangeException("error"));
+        when(mockRLM.read(any(RemoteStorageFetchInfo.class))).thenThrow(new RuntimeException("error"));
```
Comment (divijvaidya) [inline core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java:87]: is this change related to this PR?

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java [hunk 4]
```diff
@@ -41,6 +44,19 @@ public String logPrefix() {
                 return "[" + Thread.currentThread().getName() + "]";
             }
         }.logger(RemoteStorageThreadPool.class);
+        KafkaMetricsGroup metricsGroup = new KafkaMetricsGroup(this.getClass());
+        metricsGroup.newGauge(metricsNamePrefix.concat("TaskQueueSize"), new Gauge<Integer>() {
+            @Override
+            public Integer value() {
+                return RemoteStorageThreadPool.this.getQueue().size();
```
Comment (divijvaidya) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java:61]: Is this call thread safe? Note that there may be two threads working with the queue data structure at the same time, the JMX thread populating this metric and this RLM thread.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java [hunk 5]
```diff
@@ -951,6 +970,10 @@ public Thread newThread(Runnable r) {
             return threadPool;
         }
 
+        public Double getIdlePercent() {
+            return 1 - (double) scheduledThreadPool.getActiveCount() / (double) scheduledThreadPool.getCorePoolSize();
```
Comment (divijvaidya) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java:992]: Please help me understand why are we using getCorePoolSize() in stead of getPoolSize() here?

**Review by divijvaidya — COMMENTED** [review]
Thank you for making this change! I left a couple of comments. Let me know what you think about them.

File: core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java [hunk 6]
```diff
@@ -40,18 +42,29 @@
...
+    BrokerTopicStats brokerTopicStats = new BrokerTopicStats();
     LogOffsetMetadata logOffsetMetadata = new LogOffsetMetadata(100);
     Records records = mock(Records.class);
 
+    @BeforeEach
+    public void setUp() {
+        KafkaYammerMetrics.defaultRegistry()
+            .allMetrics()
+            .keySet()
+            .forEach(metricName -> KafkaYammerMetrics.defaultRegistry().removeMetric(metricName));
```
Comment (divijvaidya) [inline core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java]: You can use `TestUtils.clearYammerMetrics()`

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 2 above)
Comment (showuon) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:280]: Good point!

**Review by showuon — COMMENTED** [review]
@abhijeetk88 , thanks for the PR. Made a pass, overall LGTM!
But I didn't see the tests for `RemoteLogReaderTaskQueueSize`, `RemoteLogReaderAvgIdlePercent`, and `RemoteLogManagerTasksAvgIdlePercent`. Could you help add them? Thanks.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java (hunk 5 above)
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java:992]: A ScheduledThreadPoolExecutor takes the `corePoolSize` in its constructor, and it creates a fixed-size pool using the corePoolSize.
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledThreadPoolExecutor.html#ScheduledThreadPoolExecutor-int-

> While this class inherits from [ThreadPoolExecutor](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadPoolExecutor.html), a few of the inherited tuning methods are not useful for it. In particular, because it acts as a fixed-sized pool using corePoolSize threads

Since corePoolSize controls the pool size for a ScheduledThreadPoolExecutor, it seems better to use it here.

File: core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java (hunk 6 above)
Comment (abhijeetk88) [reply core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java]: Thanks, will change this.

File: core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java (hunk 3 above)
Comment (abhijeetk88) [reply core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java:87]: We do not capture the failed remote request metric in the case of OffsetOutOfRangeException.

I could add another test where it throws a non-OffsetOutOfRangeException keeping this test as is, but this test does nothing specific to OffsetOutOfRangeException (only requires an unsuccessful read), hence I decided to change the exception type and verify the metrics in this test itself.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java (hunk 4 above)
Comment (abhijeetk88) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java:61]: Do we need that kind of accuracy for this metric? A bit of staleness should be okay, right?

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 2 above)
Comment (abhijeetk88) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:280]: This requires a significant amount of changes. We need to know in this class if remote storage is enabled and for that to happen BrokerTopicStats needs to know that. There are many instances of object creation for BrokerTopicStats all of which will need to supply whether remote storage is enabled.

Should we address this in a separate PR? Let me know if you have better ideas to do this.

**Comment by abhijeetk88** [comment]
@showuon adding tests for those gauge metrics requires significant changes. We will need to obtain a handle for the underlying threadpool, submit tasks to the threadpool, make sure they don't progress, and then verify the gauges for non-zero values.

Shall we add those tests in a follow-up PR? Let me know if you have better ideas for these tests.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java (hunk 1 above)
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java]: Closed the metrics and added tests.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 2 above)
Comment (showuon) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:280]: Fine. Please open a JIRA ticket for this improvement. Thanks.

**Comment by satishd** [comment]
@abhijeetk88 @showuon Let us have a followup JIRA/PR on adding more test for the metrics.

**Comment by showuon** [comment]
> @abhijeetk88 @showuon Let us have a followup JIRA/PR on adding more test for the metrics.

Sure, please open a JIRA ticket to track it. Thanks.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 7]
```diff
@@ -277,6 +277,12 @@ class BrokerTopicMetrics(name: Option[String]) {
     BrokerTopicStats.TotalFetchRequestsPerSec -> MeterWrapper(BrokerTopicStats.TotalFetchRequestsPerSec, "requests"),
     BrokerTopicStats.FetchMessageConversionsPerSec -> MeterWrapper(BrokerTopicStats.FetchMessageConversionsPerSec, "requests"),
     BrokerTopicStats.ProduceMessageConversionsPerSec -> MeterWrapper(BrokerTopicStats.ProduceMessageConversionsPerSec, "requests"),
+    BrokerTopicStats.RemoteCopyBytesPerSec -> MeterWrapper(BrokerTopicStats.RemoteCopyBytesPerSec, "bytes"),
+    BrokerTopicStats.RemoteFetchBytesPerSec -> MeterWrapper(BrokerTopicStats.RemoteFetchBytesPerSec, "bytes"),
+    BrokerTopicStats.RemoteReadRequestsPerSec -> MeterWrapper(BrokerTopicStats.RemoteReadRequestsPerSec, "requests"),
+    BrokerTopicStats.RemoteWriteRequestsPerSec -> MeterWrapper(BrokerTopicStats.RemoteWriteRequestsPerSec, "requests"),
+    BrokerTopicStats.FailedRemoteReadRequestsPerSec -> MeterWrapper(BrokerTopicStats.FailedRemoteReadRequestsPerSec, "requests"),
+    BrokerTopicStats.FailedRemoteWriteRequestsPerSec -> MeterWrapper(BrokerTopicStats.FailedRemoteWriteRequestsPerSec, "requests"),
```
Comment (jeqo) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala:285]: few more naming suggestions:

```suggestion
    BrokerTopicStats.RemoteFetchRequestsPerSec -> MeterWrapper(BrokerTopicStats.RemoteFetchRequestsPerSec, "requests"),
    BrokerTopicStats.RemoteCopyRequestsPerSec -> MeterWrapper(BrokerTopicStats.RemoteCopyRequestsPerSec, "requests"),
    BrokerTopicStats.FailedRemoteCopyRequestsPerSec -> MeterWrapper(BrokerTopicStats.FailedRemoteFetchRequestsPerSec, "requests"),
    BrokerTopicStats.FailedRemoteCopyRequestsPerSec -> MeterWrapper(BrokerTopicStats.FailedRemoteCopyRequestsPerSec, "requests"),
```

**Comment by showuon** [comment]
> > @abhijeetk88 @showuon Let us have a followup JIRA/PR on adding more test for the metrics.
> 
> Sure, please open a JIRA ticket to track it. Thanks.

[KAFKA-15176](https://issues.apache.org/jira/browse/KAFKA-15176) is opened.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 7 above)
Comment (divijvaidya) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:285]: Note that the KIP version was voted with the earlier names and now changing the KIP would be equivalent to breaking the agreed contract when the KIP was accepted. I will propose to keep the metric names as they were in the KIP. If we want to introduce new ones, we can start a new KIP which deprecates these metrics and uses new ones.

Comment (jeqo) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:285]: I was suspecting this would be the case. Though as metrics haven't been implemented yet, would it be possible to amend the KIP and ask for feedback?

Comment (divijvaidya) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:285]: We at (Amazon MSK) have an implementation against the KIP-405 interfaces [1] and we explicitly tell the customers that the metric such as `RemoteBytesInPerSec` are as per KIP-405 contract. If we change the KIP now, it would be a break in contract for our customers. Hence, I would not be in favour of amending the accepted KIP at this time.

[1] see: RemoteBytesInPerSec at https://docs.aws.amazon.com/msk/latest/developerguide/metrics-details.html

**Review by showuon — APPROVED** [review]
LGTM! @abhijeetk88 , could you resolve the conflict?
@divijvaidya @jeqo , any other comments to this PR?

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 7 above)
Comment (satishd) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:285]: We should avoid having multiple versions of metrics here. As it is not yet released in Apache, we should have better naming agreeable for the majority. We do not need to have two versions here when it is released in Apache. It avoids confusion for users too. We can discuss whether that can be done by updating KIP-405(I guess I did update d a couple of them) or we can have a minor KIP for them.

@divijvaidya  This small change can be handled by organizations that adopted KIP-405.

**Review by satishd — COMMENTED** [review]
Thanks @abhijeetk88 for the PR. Overall LGTM. But we can have better naming for these metrics. That can be done in a followup PR once we decide on the names.

**Review by jeqo — APPROVED** [review]
LGTM, thanks @abhijeetk88 !

**Review by satishd — APPROVED** [review]

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 2 above)
Comment (divijvaidya) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:280]: I wasn't able to find the JIRA, so when ahead created on https://issues.apache.org/jira/browse/KAFKA-15189 . This JIRA is a blocker for 3.6 launch otherwise it will leak to regression for customers not using the Remote Storage feature,

File: core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java (hunk 3 above)
Comment (divijvaidya) [reply core/src/test/java/kafka/log/remote/RemoteLogReaderTest.java:87]: ok

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java [hunk 8]
```diff
@@ -70,4 +91,9 @@ public Thread newThread(Runnable r) {
         }
 
     }
+
+    public void removeMetrics() {
```
Comment (hudeqi) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java:100]: Hi, I am planning to clean up all the leaked metrics in the current base code recently, so I will pay close attention to the "metric", Is it possible to follow this template [here](https://github.com/apache/kafka/pull/13962)? It is better to put the metric name in a collection to prevent others from forgetting to remove when adding a new metric. Otherwise, my work in metric cleaning this area may be endless. @abhijeetk88

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java [hunk 9]
```diff
@@ -894,6 +920,7 @@ public void close() {
                 Utils.closeQuietly(indexCache, "RemoteIndexCache");
 
                 rlmScheduledThreadPool.close();
+                removeMetrics();
```
Comment (divijvaidya) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java]: This should probably be done in a try/finally after the thread pool (in the next line) has been shutdown.

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java [hunk 10]
```diff
@@ -589,6 +687,47 @@ void testIdempotentClose() throws IOException {
...
+                }
+            };
+            // Close RemoteLogManager so that metrics are removed
+            remoteLogManager.close();
+
+            KafkaMetricsGroup mockRlmMetricsGroup = mockMetricsGroupCtor.constructed().get(0);
+            KafkaMetricsGroup mockThreadPoolMetricsGroup = mockMetricsGroupCtor.constructed().get(1);
+
+            List<String> remoteLogManagerMetricNames = Collections.singletonList(REMOTE_LOG_MANAGER_TASKS_AVG_IDLE_PERCENT);
+            List<String> remoteStorageThreadPoolMetricNames = Arrays.asList(
```
Comment (divijvaidya) [inline core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java]: When we add a metric to the RemoteLogManager or to the thread pool, we will have to update this test. Alternatively, may I suggest, the pattern we have used at other places, i.e. create a list of metrics in the class where metric group is present, and verify invocations in the test for all members of that class. With this pattern, you won't have to modify the test at all when adding new metrics.

As an example, you can see  https://github.com/apache/kafka/blob/b3ce2e54f40f2d1e287d8bfd196dc5dcdbd2046d/core/src/main/scala/kafka/log/LogCleaner.scala#L529 

Separately, I also like the pattern of metric decoupling introduced used by `QuorumControllerMetrics`. It neatly encapsulates all QuorumController metrics at one place and we can potentially do similar for RemoteLogManager. This is a suggestion and feel free to ignore this.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java (hunk 4 above)
Comment (divijvaidya) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java:61]: More than accuracy, I was concerned about the impact of having two different threads accessing a non-thread safe data structure. In some cases, it could leave the structure in an inconsistent state. However, unlike complex structures like Map, seems like this queue's implementation is simple and size() just does a `AtomicInteger.get`. 

Hence, my concern is mitigated. Please consider this comment resolved.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java (hunk 5 above)
Comment (divijvaidya) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java:992]: Thank you for the explanation.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala [hunk 11]
```diff
@@ -277,6 +277,12 @@ class BrokerTopicMetrics(name: Option[String]) {
     BrokerTopicStats.TotalFetchRequestsPerSec -> MeterWrapper(BrokerTopicStats.TotalFetchRequestsPerSec, "requests"),
     BrokerTopicStats.FetchMessageConversionsPerSec -> MeterWrapper(BrokerTopicStats.FetchMessageConversionsPerSec, "requests"),
     BrokerTopicStats.ProduceMessageConversionsPerSec -> MeterWrapper(BrokerTopicStats.ProduceMessageConversionsPerSec, "requests"),
+    BrokerTopicStats.RemoteCopyBytesPerSec -> MeterWrapper(BrokerTopicStats.RemoteCopyBytesPerSec, "bytes"),
```
Comment (divijvaidya) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Please note that for some other metrics, we store aggregated topics stat using `allTopicsStats`. Are we intentionally not add remote* metrics to it?

**Review by divijvaidya — COMMENTED** [review]
Thank you for you patience so far Abhijeet. I have one last round of questions/comments. I promise to provide rapid review replies (on weekdays) to wrap up this PR soon.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java (hunk 8 above)
Comment (abhijeetk88) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteStorageThreadPool.java:100]: Addressed.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala (hunk 11 above)
Comment (abhijeetk88) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: I have added those now.

File: core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java (hunk 10 above)
Comment (abhijeetk88) [reply core/src/test/java/kafka/log/remote/RemoteLogManagerTest.java]: Addressed.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java (hunk 9 above)
Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java]: Done.

**Comment by abhijeetk88** [comment]
Addressed all the comments. Please take a look.

**Comment by showuon** [comment]
@abhijeetk88 , there are errors from spotbugs. Could you take a look?
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-13944/7/pipeline

**Review by divijvaidya — APPROVED** [review]
Thank you for your patience through this PR @abhijeetk88 (and congratulations of your first contribution to Apache Kafka!)

The changes look good to me. I am waiting for CI to reach an acceptable state now.

**Review by kamalcph — APPROVED** [review]
LGTM. Thanks for the PR!

File: core/src/main/java/kafka/log/remote/RemoteLogReader.java [hunk 12]
```diff
@@ -54,10 +60,14 @@ public Void call() {
             logger.debug("Reading records from remote storage for topic partition {}", fetchInfo.topicPartition);
 
             FetchDataInfo fetchDataInfo = rlm.read(fetchInfo);
+            brokerTopicStats.topicStats(fetchInfo.topicPartition.topic()).remoteBytesInRate().mark(fetchDataInfo.records.sizeInBytes());
+            brokerTopicStats.allTopicsStats().remoteBytesInRate().mark(fetchDataInfo.records.sizeInBytes());
             result = new RemoteLogReadResult(Optional.of(fetchDataInfo), Optional.empty());
         } catch (OffsetOutOfRangeException e) {
             result = new RemoteLogReadResult(Optional.empty(), Optional.of(e));
```
Comment (Staniel) [inline core/src/main/java/kafka/log/remote/RemoteLogReader.java:67]: Hi, came across this PR and have a qq, why we don't include the OffsetOutOfRangeException result into the failed remote read request rate metrics? If for some reason we can not, can we have a separate metrics to capture that error in metrics?

Comment (abhijeetk88) [reply core/src/main/java/kafka/log/remote/RemoteLogReader.java:67]: We do not want to include OffsetOutOfRangeException because it is not really a failure in remote reads.

We can have a separate discussion on a different metric since that change is unrelated to this PR.

**Comment by divijvaidya** [comment]
This has 33 failures, we are currently have 8-9 flaky test failures in other PRs on trunk. i am re-running the tests.

**Comment by satishd** [comment]
After the jenkins rebuild again, there are ~7 test failures that are unrelated to this PR. Merging it to trunk.

File: core/src/main/java/kafka/log/remote/RemoteLogReader.java (hunk 12 above)
Comment (Staniel) [reply core/src/main/java/kafka/log/remote/RemoteLogReader.java:67]: Thank you. I will create a JIRA to track that discussion.

Comment (Staniel) [reply core/src/main/java/kafka/log/remote/RemoteLogReader.java:67]: https://issues.apache.org/jira/browse/KAFKA-15214
